server/infinispan/src/main/java/org/hawkular/apm/server/infinispan/InfinispanPropertyCache.java

```diff
@@ -0,0 +1,225 @@
+        if (cacheContainer != null) {
+            propertyCache.startBatch();
+        }
+
+        for (Property<?> property: properties) {
+            store(tenantId, property);
+        }
+
+        if (cacheContainer != null) {
+            propertyCache.endBatch(true);
+        }
+    }
+
+    @Override
+    public void store(String tenantId, Property<?> property) throws CacheException {
+        if (log.isLoggable(Level.FINEST)) {
+            log.finest("Store property [id=" + property.getId() + "]: " + property);
+        }
+
+        propertyCache.put(property.getId(), property);
+    }
+
+    @Override
+    public void remove(String tenantId, String id) {
+        propertyCache.remove(id);
+    }
+
+    @Override
+    public void addObserver(String tenantId, String id, Observer observer) {
+        List<Observer> observers = eventListener.observersMap.get(id);
+        if (observers == null) {
+            observers = new CopyOnWriteArrayList<>();
+            eventListener.observersMap.put(id, observers);
+        }
+
+        observers.add(observer);
+    }
+
+    @Override
+    public void removeObserver(String tenantId, String id, Observer observer) {
+        List<Observer> observers = eventListener.observersMap.get(id);
+        if (observers != null) {
+            observers.remove(observer);
+
+            if (observers.isEmpty()) {
+                eventListener.observersMap.remove(id);
+            }
+        }
+    }
+
+    @Listener
+    private static class CacheEventListener {
+
+        private Map<String, List<Observer>> observersMap = new ConcurrentHashMap<>();
+
+        @CacheEntryCreated
+        public void created(CacheEntryCreatedEvent<String, Property<?>> event) {
+            if (event.isPre()) {
+                return;
+            }
+
+            List<Observer> observers = this.observersMap.get(event.getKey());
+            if (observers != null) {
+                observers.forEach(observer -> observer.created(event.getValue()));
+            }
+        }
+
+        @CacheEntryModified
+        public void updated(CacheEntryModifiedEvent<String, Property<?>> event) {
+            if (event.isPre()) {
+                return;
+            }
+
+            Property<?> oldValue = (Property<?>) ((EventImpl) event).getOldValue();
+            Property<?> newValue = event.getValue();
+
+            if (!newValue.equals(oldValue)) {
+                List<Observer> observers = this.observersMap.get(event.getKey());
+                if (observers != null) {
+                    observers.forEach(observer -> observer.updated(newValue));
+                }
+            }
+        }
+
+        @CacheEntryRemoved
+        public void removed(CacheEntryRemovedEvent<String, Property<?>> event) {
+            if (event.isPre()) {
+                return;
+            }
+
+            List<Observer> observers = this.observersMap.get(event.getKey());
+            if (observers != null && event.getOldValue() != null) {
```

Comment:
@objectiser  please look at this:

If a property with a null value gets removed observers are not modified.